src/services/getEditsForFileRename.ts

```diff
@@ -1,67 +1,143 @@
+                // Update imports in a moved file.
+                updateImportsWorker(sourceFile, changeTracker, referenceText => internalPathUpdater(sourceFile, referenceText), importText => internalPathUpdater(sourceFile, importText));
+            }
+            else {
+                // This is not a moved file, but may reference a moved file.
+                updateImportsWorker(sourceFile, changeTracker,
+                    referenceText => pathUpdater(resolveTripleslashReference(referenceText, sourceFile.fileName), referenceText, /*isImport*/ false),
+                    importText => {
+                        const resolved = host.resolveModuleNames
+                            ? host.getResolvedModuleWithFailedLookupLocationsFromCache && host.getResolvedModuleWithFailedLookupLocationsFromCache(importText, sourceFile.fileName)
+                            : program.getResolvedModuleWithFailedLookupLocationsFromCache(importText, sourceFile.fileName);
+                        return resolved && firstDefined(resolved.resolvedModule ? [resolved.resolvedModule.resolvedFileName] : resolved.failedLookupLocations, path => pathUpdater(path, importText, /*isImport*/ true));
+                    });
             }
         }
-        return result;
     }
 
-    function getPathUpdater(oldFilePath: string, newFilePath: string, host: LanguageServiceHost): (oldPath: string) => string | undefined {
+    function updateImportsWorker(sourceFile: SourceFile, changeTracker: textChanges.ChangeTracker, updateRef: (r: string) => string | undefined, updateImport: (importText: string) => string | undefined) {
+        for (const ref of sourceFile.referencedFiles) {
+            const updated = updateRef(ref.fileName);
+            if (updated !== undefined) changeTracker.replaceRangeWithText(sourceFile, ref, updated);
+        }
+
+        for (const importStringLiteral of sourceFile.imports) {
+            const updated = updateImport(importStringLiteral.text);
+            if (updated !== undefined) changeTracker.replaceRangeWithText(sourceFile, createStringRange(importStringLiteral, sourceFile), updated);
+        }
+    }
+
+    // Path updater for imports in the file(s) being moved
+    type MovedFilePathUpdater = (sourceFile: SourceFile, importText: string) => string | undefined;
+    function getMovedFilePathUpdater(oldFileOrDirPath: string, newFileOrDirPath: string): MovedFilePathUpdater {
+        const relativeNewDirToOldDir = getRelativePathFromDirectory(toDirectory(newFileOrDirPath), toDirectory(oldFileOrDirPath), /*ignoreCase*/ false);
+
+        if (relativeNewDirToOldDir === ".") return () => undefined;
+
+        return (sourceFile, importText) =>
+            !pathIsRelative(importText) ||
+                importIsInternalToDirectory(oldFileOrDirPath, sourceFile.fileName, importText) ||
+                importIsInternalToDirectory(newFileOrDirPath, sourceFile.fileName, importText)

```

Comment:
If import text is relative and it was relative path resolving in newDirectory, you need to change the text right?
"/a/b/" to "/a/c" and my import text was "in/a/b/file1" was referring to "../c/file3" you want to change it to "./file3" ?
Would it be better to use ```getRelativePathToDirectoryOrUrl``` with file and reference  new locations instead ?

Reply:
If `/a/b` is being renamed to `/a/c`, `/a/c` must not already exist -- so an import to `../c/file3` would have been an error before, right?

Reply:
I think I see what you mean in the file case, see latest commit.